### ISSUE TITLE
Input and Output names are the same `:BAM_SORT_SAMTOOLS:SAMTOOLS_SORT`

### ISSUE THREAD

**Issue by BarryDigby** [nf-core/smrnaseq#175]
### Description of the bug

Workflow exited due to line 23 in the `SAMTOOLS_SORT` process, part of the `BAM_SORT_SAMTOOLS` subworkflow:

```console
Aug-26 22:25:51.660 [Actor Thread 75] ERROR nextflow.processor.TaskProcessor - Error executing process > 'NFCORE_SMRNASEQ:SMRNASEQ:GENOME_QUANT:BAM_SORT_SAMTOOLS:SAMTOOLS_SORT (Clone1_N2_mature_hairpin_genome)'

Caused by:
  Input and output names are the same, use "task.ext.prefix" to disambiguate!

Source block:
  def args = task.ext.args ?: ''
  def prefix = task.ext.prefix ?: "${meta.id}"
  if ("$bam" == "${prefix}.bam") error "Input and output names are the same, use \"task.ext.prefix\" to disambiguate!"
  """
      samtools sort $args -@ $task.cpus -o ${prefix}.bam -T $prefix $bam
      cat <<-END_VERSIONS > versions.yml
      "${task.process}":
          samtools: \$(echo \$(samtools --version 2>&1) | sed 's/^.*samtools //; s/Using.*\$//')
      END_VERSIONS
      """

Tip: you can replicate the issue by changing to the process work dir and entering the command `bash .command.run`
```
***
Soruce:

https://github.com/nf-core/smrnaseq/blob/d1299b132a910d8adb38a5445d0c223051e14264/modules/nf-core/modules/samtools/sort/main.nf#L20-L30
***
Corresponding `modules.config` logic:

https://github.com/nf-core/smrnaseq/blob/d1299b132a910d8adb38a5445d0c223051e14264/conf/modules.config#L188-L195

***

As far as I can tell the workflow did not create a work directory for this process so I can't check out the `.command.sh` file to see what was passed to the script. I don't know enough about the workflow to understand the config logic implemented here: 

https://github.com/nf-core/smrnaseq/blob/d1299b132a910d8adb38a5445d0c223051e14264/conf/modules.config#L182-L186

Was this skipped and thus the `task.ext.prefix` was never implemented?

### Command used and terminal output

```console
`nextflow -bg run nf-core/smrnaseq -profile singularity,lugh --input 'samplesheet.csv' --outdir 'trim/' -c trim.config`
```

### Relevant files

[samplesheet.csv](https://github.com/nf-core/smrnaseq/files/9436672/samplesheet.csv)
[nextflow.log](https://github.com/nf-core/smrnaseq/files/9436676/nextflow.log)
[trim.config.txt](https://github.com/nf-core/smrnaseq/files/9436681/trim.config.txt)
Cluster config: https://github.com/nf-core/configs/blob/master/conf/lugh.config

### System information

* nextflow version 22.04.5.5708
* HPC
* SLURM
* Singularity 3.4.1
* Linux
* nf-core/smrnaseq v2.0.0

**Comment by asalhab** [comment]
I had the exact issue with the offline version. I posted my issue on slack https://nfcore.slack.com/archives/CL66GAJBF/p1660635971409769

**Comment by lpantano** [comment]
can you share your input file? I am think that maybe happen in some cases depending on the name of the samples.

**Comment by lpantano** [comment]
I just saw the file, I will take a look now

**Comment by lpantano** [comment]
Can you try with 2 samples only? if you get the same error and you can share the samples I can try to debug further.

**Comment by BarryDigby** [comment]
Sure thing. Same outcome: [2sample_log.txt](https://github.com/nf-core/smrnaseq/files/9464264/2sample_log.txt)

Here is a link to my Drive with the two samples used: https://drive.google.com/drive/folders/1R_Lx2AZSq72wno2Il0K8rdFVUXExo0bc?usp=sharing

You can recapitulate my run using the exact same files and commands listed above, just subset `samplesheet.csv` to contain only the final two lines (`Control_N2` and `Control_N3`).

**Comment by lpantano** [comment]
thank you, trying to reproduce now and see how to fix it.

**Comment by lpantano** [comment]
Hi,

for some reason I cannot reproduce. I did the following to speed up the test:

```
./nextflow run nf-core/smrnaseq -r 2.0.0  -profile docker --input 'samples.csv' --outdir tmp -c trim.config.txt  --fasta /shared/resources/grch38/hg38.fa -resume -qs 2 --skip_qc --skip_fastqc --skip_mirdeep --skip_multiqc
```

```
[1e/cdef83] process > NFCORE_SMRNASEQ:SMRNASEQ:GENOME_QUANT:INDEX_GENOME (hg38.fa)                                                                   [100%] 1 of 1, cached: 1 ✔
[0d/3f7b2a] process > NFCORE_SMRNASEQ:SMRNASEQ:GENOME_QUANT:BOWTIE_MAP_GENOME (Clone1_N3_mature_hairpin_genome)                                      [100%] 2 of 2 ✔
[6a/a71ade] process > NFCORE_SMRNASEQ:SMRNASEQ:GENOME_QUANT:BAM_SORT_SAMTOOLS:SAMTOOLS_SORT (Clone1_N3_mature_hairpin_genome)                        [100%] 2 of 2 ✔
[2e/43eb70] process > NFCORE_SMRNASEQ:SMRNASEQ:GENOME_QUANT:BAM_SORT_SAMTOOLS:SAMTOOLS_INDEX (Clone1_N3_mature_hairpin_genome)                       [100%] 2 of 2 ✔
[-        ] process > NFCORE_SMRNASEQ:SMRNASEQ:GENOME_QUANT:BAM_SORT_SAMTOOLS:BAM_STATS_SAMTOOLS:SAMTOOLS_STATS                                      -
[1a/0a38a0] process > NFCORE_SMRNASEQ:SMRNASEQ:GENOME_QUANT:BAM_SORT_SAMTOOLS:BAM_STATS_SAMTOOLS:SAMTOOLS_FLAGSTAT (Clone1_N2_mature_hairpin_genome) [100%] 2 of 2 ✔
[8c/fdc14b] process > NFCORE_SMRNASEQ:SMRNASEQ:GENOME_QUANT:BAM_SORT_SAMTOOLS:BAM_STATS_SAMTOOLS:SAMTOOLS_IDXSTATS (Clone1_N2_mature_hairpin_genome) [100%] 2 of 2 ✔
[73/469d8f] process > NFCORE_SMRNASEQ:SMRNASEQ:CUSTOM_DUMPSOFTWAREVERSIONS (1)                                                                       [100%] 1 of 1 ✔
-[nf-core/smrnaseq] Pipeline completed successfully-
Completed at: 31-Aug-2022 22:42:03
Duration    : 1h 21m 39s
CPU hours   : 8.5 (58.5% cached)
Succeeded   : 45
Cached      : 8

```

My nextflow version is: version 22.04.5 build 5708

This is the samples.csv:

```
sample,fastq_1
Clone1_N2,LNCaP/Ctl-N2-R1_S2_L001_R1_001.fastq.gz
Clone1_N3,LNCaP/Ctl-N3-R1_S3_L001_R1_001.fastq.gz
```

config file:

```
params{

    // Reference files
    mirtrace_species           = 'hsa'

    // Trimming options
    clip_r1                    = 0
    three_prime_clip_r1        = 0
    three_prime_adapter        = "TGGAATTCTCGGGTGCCAAGG"
    min_length                 = 17
    skip_qc                    = false
    skip_fastqc                = false
    skip_multiqc               = false
    skip_mirdeep               = false
    skip_trimming              = false
    save_reference             = false
    trim_galore_max_length     = 40
}
```

I had to change the genome source because in my env I cannot use the genomes from nf-core S3 location, but that shouldn't change the behavior. Do you mind to try this (with the same nextflow veresion and the same location of the input file) and see what happens?

**Comment by BarryDigby** [comment]
Ok so the following ran to completion, identical to your setup detailed above:

```console
nextflow run nf-core/smrnaseq -r 2.0.0 -profile singularity --input '2samps.csv' --outdir tmp -c trim.config --fasta grch37.fa -qs 2 --skip_qc --skip_mirdeep --skip_multiqc
```

```console
executor >  slurm (53)
[fe/ff3403] process > NFCORE_SMRNASEQ:SMRNASEQ:INPUT_CHECK:SAMPLESHEET_CHECK (2samps.csv)                                                             [100%] 1 of 1 ✔
[-        ] process > NFCORE_SMRNASEQ:SMRNASEQ:CAT_FASTQ                                                                                              -
[73/ded4c7] process > NFCORE_SMRNASEQ:SMRNASEQ:MIRTRACE:MIRTRACE_RUN                                                                                  [100%] 1 of 1 ✔
[e6/4788c8] process > NFCORE_SMRNASEQ:SMRNASEQ:FASTQC_TRIMGALORE:TRIMGALORE (Control_N3)                                                              [100%] 2 of 2 ✔
[07/d0a30d] process > NFCORE_SMRNASEQ:SMRNASEQ:MIRNA_QUANT:PARSE_MATURE                                                                               [100%] 1 of 1 ✔
[d3/b33b5f] process > NFCORE_SMRNASEQ:SMRNASEQ:MIRNA_QUANT:FORMAT_MATURE (mature.fa_igenome.fa)                                                       [100%] 1 of 1 ✔
[51/e2d164] process > NFCORE_SMRNASEQ:SMRNASEQ:MIRNA_QUANT:PARSE_HAIRPIN                                                                              [100%] 1 of 1 ✔
[4c/225a6a] process > NFCORE_SMRNASEQ:SMRNASEQ:MIRNA_QUANT:FORMAT_HAIRPIN (hairpin.fa_igenome.fa)                                                     [100%] 1 of 1 ✔
[67/0d5b2b] process > NFCORE_SMRNASEQ:SMRNASEQ:MIRNA_QUANT:INDEX_MATURE                                                                               [100%] 1 of 1 ✔
[cc/ee037d] process > NFCORE_SMRNASEQ:SMRNASEQ:MIRNA_QUANT:INDEX_HAIRPIN                                                                              [100%] 1 of 1 ✔
[f2/97df49] process > NFCORE_SMRNASEQ:SMRNASEQ:MIRNA_QUANT:BOWTIE_MAP_MATURE (Control_N3_mature)                                                      [100%] 2 of 2 ✔
[a1/186402] process > NFCORE_SMRNASEQ:SMRNASEQ:MIRNA_QUANT:BOWTIE_MAP_HAIRPIN (Control_N3_mature_hairpin)                                             [100%] 2 of 2 ✔
[1d/7bd8e4] process > NFCORE_SMRNASEQ:SMRNASEQ:MIRNA_QUANT:BAM_STATS_MATURE:SAMTOOLS_SORT (Control_N3_mature)                                         [100%] 2 of 2 ✔
[bb/87cafd] process > NFCORE_SMRNASEQ:SMRNASEQ:MIRNA_QUANT:BAM_STATS_MATURE:SAMTOOLS_INDEX (Control_N3_mature)                                        [100%] 2 of 2 ✔
[9c/986f8a] process > NFCORE_SMRNASEQ:SMRNASEQ:MIRNA_QUANT:BAM_STATS_MATURE:BAM_STATS_SAMTOOLS:SAMTOOLS_STATS (Control_N2_mature)                     [100%] 2 of 2 ✔
[8a/0fb83d] process > NFCORE_SMRNASEQ:SMRNASEQ:MIRNA_QUANT:BAM_STATS_MATURE:BAM_STATS_SAMTOOLS:SAMTOOLS_FLAGSTAT (Control_N2_mature)                  [100%] 2 of 2 ✔
[39/6eb9c2] process > NFCORE_SMRNASEQ:SMRNASEQ:MIRNA_QUANT:BAM_STATS_MATURE:BAM_STATS_SAMTOOLS:SAMTOOLS_IDXSTATS (Control_N3_mature)                  [100%] 2 of 2 ✔
[87/00478c] process > NFCORE_SMRNASEQ:SMRNASEQ:MIRNA_QUANT:BAM_STATS_HAIRPIN:SAMTOOLS_SORT (Control_N3_mature_hairpin)                                [100%] 2 of 2 ✔
[a6/5fe345] process > NFCORE_SMRNASEQ:SMRNASEQ:MIRNA_QUANT:BAM_STATS_HAIRPIN:SAMTOOLS_INDEX (Control_N3_mature_hairpin)                               [100%] 2 of 2 ✔
[e9/5e2e78] process > NFCORE_SMRNASEQ:SMRNASEQ:MIRNA_QUANT:BAM_STATS_HAIRPIN:BAM_STATS_SAMTOOLS:SAMTOOLS_STATS (Control_N2_mature_hairpin)            [100%] 2 of 2 ✔
[ae/ee71f2] process > NFCORE_SMRNASEQ:SMRNASEQ:MIRNA_QUANT:BAM_STATS_HAIRPIN:BAM_STATS_SAMTOOLS:SAMTOOLS_FLAGSTAT (Control_N3_mature_hairpin)         [100%] 2 of 2 ✔
[07/f26727] process > NFCORE_SMRNASEQ:SMRNASEQ:MIRNA_QUANT:BAM_STATS_HAIRPIN:BAM_STATS_SAMTOOLS:SAMTOOLS_IDXSTATS (Control_N3_mature_hairpin)         [100%] 2 of 2 ✔
[7e/efcb65] process > NFCORE_SMRNASEQ:SMRNASEQ:MIRNA_QUANT:EDGER_QC                                                                                   [100%] 1 of 1 ✔
[7b/9ac4c4] process > NFCORE_SMRNASEQ:SMRNASEQ:MIRNA_QUANT:SEQCLUSTER_SEQUENCES (Control_N3_seqcluster)                                               [100%] 2 of 2 ✔
[42/d6886f] process > NFCORE_SMRNASEQ:SMRNASEQ:MIRNA_QUANT:BOWTIE_MAP_SEQCLUSTER (Control_N2_seqcluster)                                              [100%] 2 of 2 ✔
[d8/22cbd4] process > NFCORE_SMRNASEQ:SMRNASEQ:MIRNA_QUANT:MIRTOP_QUANT                                                                               [100%] 1 of 1 ✔
[03/c5d700] process > NFCORE_SMRNASEQ:SMRNASEQ:MIRNA_QUANT:TABLE_MERGE                                                                                [100%] 1 of 1 ✔
[96/c709a5] process > NFCORE_SMRNASEQ:SMRNASEQ:GENOME_QUANT:INDEX_GENOME (grch37.fa)                                                                  [100%] 1 of 1 ✔
[38/4a7d42] process > NFCORE_SMRNASEQ:SMRNASEQ:GENOME_QUANT:BOWTIE_MAP_GENOME (Control_N3_mature_hairpin_genome)                                      [100%] 2 of 2 ✔
[d9/da9f7b] process > NFCORE_SMRNASEQ:SMRNASEQ:GENOME_QUANT:BAM_SORT_SAMTOOLS:SAMTOOLS_SORT (Control_N3_mature_hairpin_genome)                        [100%] 2 of 2 ✔
[91/b4f002] process > NFCORE_SMRNASEQ:SMRNASEQ:GENOME_QUANT:BAM_SORT_SAMTOOLS:SAMTOOLS_INDEX (Control_N3_mature_hairpin_genome)                       [100%] 2 of 2 ✔
[-        ] process > NFCORE_SMRNASEQ:SMRNASEQ:GENOME_QUANT:BAM_SORT_SAMTOOLS:BAM_STATS_SAMTOOLS:SAMTOOLS_STATS                                       -
[ed/4dd1ca] process > NFCORE_SMRNASEQ:SMRNASEQ:GENOME_QUANT:BAM_SORT_SAMTOOLS:BAM_STATS_SAMTOOLS:SAMTOOLS_FLAGSTAT (Control_N2_mature_hairpin_genome) [100%] 2 of 2 ✔
[05/086f2c] process > NFCORE_SMRNASEQ:SMRNASEQ:GENOME_QUANT:BAM_SORT_SAMTOOLS:BAM_STATS_SAMTOOLS:SAMTOOLS_IDXSTATS (Control_N3_mature_hairpin_genome) [100%] 2 of 2 ✔
[79/5a1654] process > NFCORE_SMRNASEQ:SMRNASEQ:CUSTOM_DUMPSOFTWAREVERSIONS (1)                                                                        [100%] 1 of 1 ✔
-[nf-core/smrnaseq] Pipeline completed successfully-
Completed at: 01-Sep-2022 19:32:35
Duration    : 30m 48s
CPU hours   : 3.7
Succeeded   : 53
```

*** 

Are we sure that the `--fasta` parameter is not differentiating factor here? I refer to my previous comment about the `modules.config` file.. perhaps iGenomes is doing something wonky (it wouldn't be the first time) :) 

(My troubleshooting aside - out of curiosity why are the lines 182-185 repeated when they exist in the main workflow script?) 

https://github.com/nf-core/smrnaseq/blob/d1299b132a910d8adb38a5445d0c223051e14264/conf/modules.config#L182-L195

I will remove these lines and the `if` statement on a local fork to see if the process runs without input/output collisions. I'll let you know how it goes. Thanks for your time on this btw 👍🏻

**Comment by lpantano** [comment]
After seeing that bit of code, maybe `fasta` is the difference. Not sure why is that there right now. But it seems to be the culprit at first sight...let us know, it is a good catch.

**Comment by BarryDigby** [comment]
@lpantano the PR fixed this current issue, at least in my specific use case.

However.. the `mirtop` MultiQC module failed due to an empty file, I can investigate the error and open a new issue if you are interested - happy to help debug rare edge cases.

**Comment by apeltzer** [comment]
The mirtop part was fixed with an update of MultiQC to v1.14 - which fixed several edge cases. So this should be fine now @BarryDigby , please give the latest dev branch a try to make sure all works for you too :-)

**Comment by BarryDigby** [comment]
@apeltzer miRTop + MultiQC v1.14 works 👍🏼 (`revision: 28a55ecefc [dev]`)

found something small, PR to come
